### PR TITLE
[CFP-61] Migrate Allocation forms to GOVUKDesignSystemFormBuilder

### PR DESCRIPTION
**Pull request by naseberry**
#### What
Migrate provider form to the GOVUK design system form builder

#### Ticket
[Migrate caseworker allocation forms to the GOVUK design system form builder](https://dsdmoj.atlassian.net/browse/CFP-61)

#### Why
Part of a larger change to:

- continue migration to GOVUK Frontend
- resolve accessibility issues identified in the 2020 DAC report
- reduce then remove deprecated code

### CODE REVIEW

**Review by jsugarman — COMMENTED** [review]
Looks good code-wise and can't see anything obviously problematic.

It would be good to extend the rather simple `allocation.feature` but in lieu of that I will need to test side by side.

**Comment by jsugarman** [comment]
Have found a problem with allocation on the new styled page when comparing side by side. 

The ability to allocate a specific number of claims to a case worker seems to have broken. It looks like it is allocating **up to 25 claims** to the case worker irrespective of how many you specify to allocate.

This will need fixing before it can go in. Would be good to have a cucumber feature test around it too.

Screenshot 2021-05-28 at 16 33 46

**Comment by jsugarman** [comment]
A smaller issue which is not related to these changes. 

When allocating claims when there are no claims of that type you get an empty error message/flash

Screenshot 2021-05-28 at 16 42 05

**Comment by jsugarman** [comment]
Another edge issue not related to this change.

When you have had an error on the page and then immediately allocate successfully the flash styling is not replaced with the green version


Screenshot 2021-05-28 at 16 46 55

**Comment by naseberry** [comment]
> Have found a problem with allocation on the new styled page when comparing side by side.
> 
> The ability to allocate a specific number of claims to a case worker seems to have broken. It looks like it is allocating **up to 25 claims** to the case worker irrespective of how many you specify to allocate.
> 
> This will need fixing before it can go in. Would be good to have a cucumber feature test around it too.
> 
> Screenshot 2021-05-28 at 16 33 46

This is now resolved. I have updated the JS selector name from `#quantity_to_allocate` to `#quantity-to-allocate-field`.

I've extended the existing feature spec to cover this.

**Comment by naseberry** [comment]
> A smaller issue which is not related to these changes.
> 
> When allocating claims when there are no claims of that type you get an empty error message/flash
> 
> Screenshot 2021-05-28 at 16 42 05

This has now been resolved. I'm checking the http response status and setting an error message.

**Comment by naseberry** [comment]
> > A smaller issue which is not related to these changes.
> > When allocating claims when there are no claims of that type you get an empty error message/flash
> > Screenshot 2021-05-28 at 16 42 05
> 
> This has now been resolved. I'm checking the http response status and setting an error message.

Resolved, we are now removing the notification modifier class and applying the correct class.

File: app/views/case_workers/admin/allocations/_allocation.html.haml [hunk 1]
```diff
@@ -1,54 +1,47 @@
...
+  = f.hidden_field :value_band_id, id: :value_band_id, value: params[:value_band_id] if params[:value_band_id]
+
+  .govuk-grid-row
+    .govuk-grid-column-two-thirds.scheme-filters
+      = render partial: 'scheme_filters', locals: { f: f }
+
+  = f.govuk_fieldset legend: { text: t('.allocate_claims'), size: 'm' } do
+    .govuk-grid-row
+      .govuk-grid-column-one-third
+        = f.govuk_number_field :quantity_to_allocate, label: { text: t('.no_of_claims') }
```
Comment (jsugarman) [inline app/views/case_workers/admin/allocations/_allocation.html.haml:22]: was this line involved in the fix to the bug for quantify allocated I found on the first review? Be good to be able to see the changes for fixes - you can also ways squash prior to merging

Comment (naseberry) [reply app/views/case_workers/admin/allocations/_allocation.html.haml:22]: Apologies, it was this line `const quantityToAllocate = $('#quantity-to-allocate-field').val() || false` in the file `app/webpack/javascripts/modules/Modules.AllocationDataTable.js` @jsugarman

File: app/webpack/javascripts/modules/Modules.AllocationDataTable.js [hunk 2]
```diff
@@ -343,16 +354,19 @@ moj.Modules.AllocationDataTable = {
...
         self.reloadScheme({
           scheme: self.searchConfig.scheme
         })
       }).error(function (data) {
-        self.ui.$notificationMsg.removeClass('govuk-!-display-none')
+        self.ui.$notificationMsg.removeClass('govuk-!-display-none govuk-notification-banner--success')
         self.ui.$notificationMsg.addClass('govuk-notification-banner--error')
+        if (data.status === 422) {
+          return self.ui.$notificationMsg.find('.govuk-notification-banner__heading').html('Unable to allocate claim')
+        }
```
Comment (naseberry) [inline app/webpack/javascripts/modules/Modules.AllocationDataTable.js:369]: Resolution to the second issue @jsugarman

File: app/webpack/javascripts/modules/Modules.AllocationDataTable.js [hunk 3]
```diff
@@ -343,16 +354,19 @@ moj.Modules.AllocationDataTable = {
           claim_ids: data
         }
       }).success(function (data) {
-        self.ui.$notificationMsg.removeClass('govuk-!-display-none')
+        self.ui.$notificationMsg.removeClass('govuk-!-display-none govuk-notification-banner--error')
```
Comment (naseberry) [inline app/webpack/javascripts/modules/Modules.AllocationDataTable.js:357]: Resolution to the third issue @jsugarman

File: app/webpack/javascripts/modules/Modules.AllocationDataTable.js [hunk 4]
```diff
@@ -343,16 +354,19 @@ moj.Modules.AllocationDataTable = {
...
         self.ui.$notificationMsg.addClass('govuk-notification-banner--success')
-        self.ui.$notificationMsg.find('.govuk-notification-banner__heading').html(data.allocated_claims.length + ' claims have been allocated to ' + $('#allocation_case_worker_id').val())
+        self.ui.$notificationMsg.find('.govuk-notification-banner__heading').html(data.allocated_claims.length + ' claims have been allocated to ' + $('#allocation-case-worker-id-field').val())
 
         self.reloadScheme({
           scheme: self.searchConfig.scheme
         })
       }).error(function (data) {
-        self.ui.$notificationMsg.removeClass('govuk-!-display-none')
+        self.ui.$notificationMsg.removeClass('govuk-!-display-none govuk-notification-banner--success')
```
Comment (naseberry) [inline app/webpack/javascripts/modules/Modules.AllocationDataTable.js:365]: Resolution to the third issue @jsugarman

File: app/webpack/javascripts/modules/Modules.AllocationDataTable.js [hunk 5]
```diff
@@ -301,15 +312,15 @@ moj.Modules.AllocationDataTable = {
...
     $('.allocation-submit').on('click', function (e) {
-      self.ui.$notificationMsg.removeClass('govuk-!-display-none')
+      self.ui.$notificationMsg.removeClass('govuk-!-display-none govuk-notification-banner--error govuk-notification-banner--success')
       self.ui.$notificationMsg.find('.govuk-notification-banner__heading').html('Allocating.. please wait a moment..')
 
       e.preventDefault()
       self.ui.$submit.prop('disabled', true)
 
-      const quantityToAllocate = $('#quantity_to_allocate').val() || false
+      const quantityToAllocate = $('#quantity-to-allocate-field').val() || false
```
Comment (jsugarman) [inline app/webpack/javascripts/modules/Modules.AllocationDataTable.js:321]: 👍 I guess this is the line that fixes the quantity allocation bug

**Review by jsugarman — APPROVED** [review]
👍 reviewed locally again and all seems to function as expected.

**Pull request merged**